### ISSUE TITLE
Replace the center-cancel vocal mute with real ML separation (BS-RoFormer)

### ISSUE THREAD

**Issue by kevinch3** [kevinch3/NicotinD#603]
## Why

The karaoke vocal mute is one ffmpeg line (`transcode.ts:73`, `pan=stereo|c0=c0-c1|c1=c1-c0`).
It removes anything panned center, so it takes the kick and bass with the vocal:
**−9.35 dB of sub-bass** on a measured real track. (It also collapses to silence in mono —
that half is #602, and is fixable independently.)

Spike run 2026-08-20: **`docs/vocal-isolation-spike.md`**.

## What the spike settled

- **Model: `anvuew/BS-RoFormer`** (GPL-3.0 → compatible with our AGPL-3.0-only), 51.0 M
  params, `num_stems: 1` (vocals only — exactly our case). Measured on a real track it
  leaves the low end intact (−1.60 dB sub-bass vs the filter's −9.35 dB).
- The obvious-looking alternative, `HiDolen/Mini-BS-RoFormer-V2-46.8M`, is
  **CC-BY-NC-4.0 — rejected on license**, and is not meaningfully smaller anyway (46.8 M
  vs 51.0 M).
- Pin **`bs-roformer==0.4.1`**; it loads the checkpoint 0-missing/0-unexpected. Later
  releases mismatch 620+ tensors.
- **ONNX export still fails**, so this cannot ride inside the Essentia sidecar today —
  but the blocker is now a *single* complex-valued `F.pad`, not Demucs' whole-STFT
  problem. Narrow enough to be worth one more attempt (spike §7, task 2).
- **CPU-only on-demand is not viable**: RTF ≈ 4.1×, 2.7 GB RSS → ~14 min of CPU per song.

## Blocking work, in order

1. **Measure on the GPU.** Not done — the spike box has no GPU. RTF + peak VRAM for
   BS-RoFormer on the P4000, against Demucs' RTF 0.10 / 2.7 GB baseline. Command in
   spike §7. The 8 GB card already hosts essentia-tensorflow (~4.3 GB) and ollama
   (~2.7 GB), so this number decides whether it is schedulable at all.
2. **Decide when separation runs** — the one open product question. Separation is not a
   streaming filter, so `?vocals=off` becomes a *precomputed stem* (the existing `novox`
   transcode-cache entry is the slot). Options: (a) on demand at first karaoke play,
   (b) an opt-in nightly enrichment task, (c) hybrid — precompute liked/recently-played,
   on-demand for the rest. Nothing below the API seam differs between them.
3. Then: sidecar shape (separate torch image per the previous spike's option B, unless
   the ONNX retry in §7 succeeds), enrichment-task wiring, gate/ledger conventions.

Refs #602 (the mono bug in the filter this would replace).

**Comment by kevinch3** [comment]
## Decision (owner, 2026-08-20): opt-in, on demand, no precompute

Step 2 of the plan above is settled. Separation runs **on demand at first karaoke play**,
behind an **opt-in** setting. **No library precompute** — no nightly enrichment task, no
liked/recently-played batch tier.

The `novox` transcode-cache entry still caches the result after the first separation.
That's caching, not precomputation: work happens only for a track someone actually asked
to sing to.

### What this makes newly in-scope

- **GPU gating.** CPU-only is RTF ≈ 4.1× → ~14 min for one song, so an on-demand path on a
  CPU box is not offerable. A deployment without a usable GPU must hide the toggle or keep
  center-cancel as an explicitly labelled degraded mode. It must never silently take the
  14-minute path.
- **A real "preparing" state** on first play, bounded by the GPU RTF.
- **The toggle must reflect availability, not just preference** — reuse the
  `configured && healthy` shape the analysis sidecar already reports on
  `GET /api/admin/review`.
- **Foreground GPU contention.** Unlike enrichment this request cannot wait, which is the
  inverse of the `gpuBusyPercent` yield rule.

### Newly blocking

**#605** — measured while setting up the GPU benchmark: the analysis sidecar reports
`{"loaded": false}` (idle release fired) yet still pins **7,626 MiB of the 8,192 MiB
card**. `ANALYSIS_IDLE_RELEASE_SEC` frees models, not memory. With ~561 MiB free there is
no room to run a separator on demand at all, so #605 is now a hard prerequisite for this
issue rather than an adjacent cleanup.

Doc updated: `docs/vocal-isolation-spike.md` §6.

**Comment by kevinch3** [comment]
## GPU measurement done — Quadro P4000, 2026-08-20

Ran in a throwaway `pytorch/pytorch:2.4.1-cuda12.1-cudnn9-runtime` container on kpc
against a real library track. Image and staged files removed afterwards.

```
GPU: Quadro P4000 | torch 2.4.1+cu121 | cc (6, 1)
model on GPU: 51.0M params
  chunk 1: 5.66s   chunk 2: 5.66s   chunk 3: 5.68s   chunk 4: 5.73s
RESULT  mean 5.68s per 21.77s chunk -> RTF 0.261x
        torch peak alloc 2082 MiB | reserved 2836 MiB | nvidia-smi total-used 3029 MiB
        -> a 3.5 min song ≈ 55s of GPU
```

**Verdict: comfortably viable on GPU.** ~2.6× slower than Demucs (RTF 0.10) at comparable
VRAM, which is the price of the quality in the spike doc §3.

### The important consequence for the on-demand decision

RTF 0.261 means separation runs **~3.8× faster than playback**. So the on-demand shape does
**not** have to be a 55-second wait: separating chunk *n+1* while chunk *n* plays lets audio
start after the **first ~5.7 s chunk** and stay ahead of the playhead for the rest of the
track. ~6 s to first audio instead of ~55 s. This is the biggest UX lever in the design and
should be assumed in the implementation, not treated as an optimisation for later.

It needs the stem written incrementally into the `novox` cache entry, with a partial entry
never mistakable for a complete one — the same discipline the transcode cache already
applies via size-in-key.

### Deployment facts this pins down

- **Pascal (cc 6.1) forces a cu121-era torch** (2.4.1 here). CUDA 13 / current torch builds
  drop `sm_61`. Not a preference — the newer stack simply will not run on this card.
- **~3.0 GB VRAM** for the separator. Alongside the analysis sidecar's ~4.3 GB that is
  ~7.3 GB of 8,192 MiB — it fits, but *only* if #605 is fixed. Ollama's 2.7 GB would not
  also fit.

### #605 re-confirmed in passing

`docker restart nicotind-analysis-1` took the card from 7,631 MiB → **85 MiB**. Within
minutes of the sidecar's first inference it was back to **7,631 MiB** and stayed there at
0 % utilisation. A container restart is the only thing that frees it.

**Comment by kevinch3** [comment]
## Re-kick 2026-09-02 — knowledge refresh + the remaining design decisions

Step 3 of the plan above ("sidecar shape, wiring, gate conventions") is now designed. Four things in the
text above are stale, and four product decisions were taken today.

### What changed since 08-20

| Topic | Above | Now |
| --- | --- | --- |
| Torch pin | "cu121-era torch (2.4.1) — a deployment fact" | PyTorch dropped Pascal (`sm_61`) only from its **CUDA 12.8+** wheels (torch 2.8+); **CUDA 12.6 wheels stay as the legacy lane** through at least torch 2.12. kpc's driver is 580.173.02 (CUDA 13-capable). Pin a current torch from the **cu126** index. |
| Loader | `bs-roformer==0.4.1` + hand-filtered config keys | Still right. `audio-separator` (MIT) bundles fork-compatible roformer code but **only loads registry checkpoints**, and `anvuew/…ft1_sdr_12.55` is not in it. `bs-roformer` upstream is 1.2.4 and still mismatches. `yaml.UnsafeLoader` is unnecessary — the only custom tag is `!!python/tuple`. |
| Models | anvuew (GPL-3.0) | Kim's **Mel-Band RoFormer was relicensed MIT (2026-04)** — a future swap candidate, unmeasured. The viperx "12.9755" model most tools default to has **no license**. anvuew also has a "karaoke" model (lead only, keeps backing vocals) — a product option, out of scope. |
| GPU | 561 MiB free | **2,235 MiB used / 8,192**, analysis sidecar only. Immich ML idle, Ollama inactive. ~5.9 GB free; separator (3.0 GB) + analysis co-host with margin. |
| #602 | filed | still open — becomes PR 1. |

### What the code ruled out

**Progressive (chunk-by-chunk) serving is off the table.** `GET /stream` is strictly
complete-then-serve: `serveFileWithRange` reads the size once, the ETag is size+mtime, iOS Safari
tail-probes with suffix ranges, and Bun serialises any unknown-length body as a chunked 206 — the
exact shape Firefox and iOS stall on (docs/library-scanner.md "Transcode cache integrity"). The
player's false-ended recovery would also fight a growing file. Serving a partial stem as one file
would mean lying about `Content-Length`; doing it properly means MSE/HLS, which iPhone PWAs lack.
The "~6 s to first audio" idea in the comment above is retired.

### Decisions (owner, 2026-09-02)

1. **Wait for the whole stem.** The separated file lands in the cache complete, through the
   existing tmp+rename + ffprobe-duration contract, and is served by the unchanged range path.
2. **Start trigger:** separate the current track **when the fullscreen karaoke overlay opens**, and
   while vocals are muted **also prepare the next queued track**. Nothing outside an active karaoke
   session — the "no library precompute" decision stands.
3. **Fallback:** the center-cancel filter stays as an explicitly labelled **basic** mode wherever
   ML is unusable (not configured, opt-in off, sidecar unhealthy, separation failed). #602 is fixed
   first so the fallback is mono-safe.
4. **During the wait:** keep playing the original mix; the toggle shows "Preparing… ~Ns" (track
   length × measured RTF); when the stem lands, swap `audio.src` at the same position (the
   mechanism the toggle already uses).

### Design in one paragraph

A `packages/separator/` sidecar mirroring `packages/analysis/` (FastAPI, single spawned worker
**with** a per-call timeout and lifespan shutdown — two gaps the analysis `ProcessRunner` has —
idle release by stopping the worker so the CUDA context returns too, checkpoint baked at build,
GPU-only by contract: `unavailable` without CUDA) returns the instrumental as FLAC. The API lands it
in a new `stem-cache/` and derives the served file through the unchanged transcode path as a `|stem`
cache variant keyed on the original's identity, so `?vocals=off` picks stem-or-basic server-side
and never blocks on the GPU. A small idempotent `POST/GET /api/stream/:id/stem` status endpoint
(`unavailable | queued | preparing | ready | failed`, with ETA) drives a web state machine
(pending mute, next-track warm-up). Opt-in is an admin toggle with `NICOTIND_SEPARATOR_URL` as the
env floor, the same shape as the acquisition kill-switch. Compose: the service lives only in
`docker-compose.gpu.yml`.

### Found in passing

The web only appends `?vocals=off` in the toggle effect; the track-load, gapless-standby and
recovery paths load without it, so **a muted karaoke session un-mutes on the next song** today.
Filed as #889.

### PRs, in order (each merged + deployed before the next)

1. `fix(streaming)`: mono-safe center-cancel (Closes #602).
2. `fix(player)`: vocal mute drops on track advance (#889).
3. `feat(separator)`: sidecar image + GPU overlay + CI/deploy — inert until 4, and the prod
   GPU/VRAM verification step.
4. `feat(karaoke)`: API + web behind the admin opt-in (Closes #603).
